dvc/tree/dropbox.py

```diff
@@ -0,0 +1,289 @@
+            config[EXPIRATION] = datetime.datetime.fromisoformat(
+                config[EXPIRATION]
+            )
+        else:
+            config[EXPIRATION] = datetime.datetime.now()
+        dbx = dropbox.Dropbox(
+            oauth2_access_token=config[ACCESS_TOKEN],
+            oauth2_access_token_expiration=config[EXPIRATION],
+            oauth2_refresh_token=config[REFRESH_TOKEN],
+            app_key=self.DROPBOX_APP_KEY,
+            app_secret=self.DROPBOX_APP_SECRET,
+        )
+        dbx.check_and_refresh_access_token()
+        dbx.check_user()
+        return Auth(dbx, config)
+
+    @abstractmethod
+    def save(self, creds):
+        raise NotImplementedError
+
+
+class CliCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return True
+
+    def _gather_auth(self):
+        import dropbox
+
+        flow = dropbox.DropboxOAuth2FlowNoRedirect(
+            consumer_key=self.DROPBOX_APP_KEY,
+            consumer_secret=self.DROPBOX_APP_SECRET,
+            token_access_type="offline",
+        )
+        authorize_url = flow.start()
+        print("Go to the following link in your browser:")
+        print()
+        print("    " + authorize_url)
+        print()
+        code = input("Enter verification code: ").strip()
+        auth = flow.finish(code)
+        return {
+            ACCESS_TOKEN: auth.access_token,
+            EXPIRATION: auth.expires_at.isoformat(),
+            REFRESH_TOKEN: auth.refresh_token,
+        }
+
+    def save(self, creds):
+        pass
+
+
+class EnvCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return bool(os.environ.get(REFRESH_TOKEN, False))
+
+    def _gather_auth(self):
+        return {
+            ACCESS_TOKEN: os.environ.get(ACCESS_TOKEN, ""),
+            EXPIRATION: os.environ.get(EXPIRATION, ""),
+            REFRESH_TOKEN: os.environ[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        os.environ.update(creds)
+
+
+class FileCredProvider(DropboxCredProvider):
+    DEFAULT_FILE = "dropbox-user-credentials.json"
+    DROPBOX_CREDENTIALS_FILE = "DROPBOX_CREDENTIALS_DATA"
+
+    def __init__(self, repo):
+        self._cred_location = os.environ.get(
+            self.DROPBOX_CREDENTIALS_FILE,
+            os.path.join(repo.tmp_dir, self.DEFAULT_FILE),
+        )
+
+    def can_auth(self):
+        if not os.path.exists(self._cred_location):
+            return False
+        with open(self._cred_location) as infile:
+            return REFRESH_TOKEN in json.load(infile)
+
+    def _gather_auth(self):
+        with open(self._cred_location) as infile:
+            config = json.load(infile)
+        return {
+            ACCESS_TOKEN: config.get(ACCESS_TOKEN, ""),
+            EXPIRATION: config.get(EXPIRATION, ""),
+            REFRESH_TOKEN: config[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        with open(self._cred_location, "w") as outfile:
+            json.dump(creds, outfile)
```

Comment:
I'm getting:

```
 File "/Users/ivan/Projects/dvc/dvc/tree/dropbox.py", line 137, in save
    json.dump(creds, outfile)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
```

on the regular CLI flow and a file that is written partially

Reply:
Good catch, thanks! I'm on it 🙂